pkg/nodestate/state_v2.go

```diff
@@ -0,0 +1,124 @@
+package nodestate
+
+import (
+	"context"
+	"fmt"
+	"github.com/benbjohnson/clock"
+	v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	"sync"
+	"time"
+)
+
+type NodeStateV2 struct {
+	scheduledPods map[string][]string
+	client        kubernetes.Interface
+	lock          *sync.RWMutex
+	clock         clock.Clock
+}
+
+func NewNodeStateV2(client kubernetes.Interface) NodeStateInterface {
+	return internalNewNodeStateV2(client, clock.New())
+}
+
+func internalNewNodeStateV2(client kubernetes.Interface, c clock.Clock) NodeStateInterface {
+	var lock = sync.RWMutex{}
+	return &NodeStateV2{
+		client:        client,
+		scheduledPods: make(map[string][]string),
+		lock:          &lock,
+		clock:         c,
+	}
+}
+
+func (n *NodeStateV2) UnhealthyPods(nodeName string) int {
+	// copied from https://github.com/kubernetes/kubernetes/blob/4f2d7b93da2464a3147e0a7e71d896dd2bade9ad/pkg/printers/internalversion/describe.go#L2451
+	fieldSelector, err := fields.ParseSelector("spec.nodeName=" + nodeName + ",status.phase!=" + string(v1.PodSucceeded) + ",status.phase!=" + string(v1.PodFailed))
+	if err != nil {
+		klog.Error("Failed to create field selector for running pod calculation with error %v", err)
+		return -1
+	}
+
+	nodeNonTerminatedPodsList, err := n.client.CoreV1().Pods("").List(context.TODO(), meta_v1.ListOptions{FieldSelector: fieldSelector.String()})
+	if err != nil {
+		klog.Error("Failed to list pods on node %s with error %v", nodeName, err)
+		return -1
+	}
+
+	unhealthyPods := 0
```

Comment:
I'd rename this variable to notReadyPods. And maybe the whole method should be renamed this way for better code readability.

Reply:
valid point should be fixed in updated commit